// @roots/blade-loader/lib/loaders/index.cjs
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.make = exports.repository = void 0;
exports.repository = {
    css: {
        pattern: /@css([\s\S]*?)@endcss/,
        extension: `.css`,
        loader: require.resolve(`./css.cjs`),
    },
    scss: {
        pattern: /@scss([\s\S]*?)@endscss/,
        extension: `.scss`,
        loader: require.resolve(`./scss.cjs`),
    },
    js: {
        pattern: /@js([\s\S]*?)@endjs/,
        extension: `.js`,
        loader: require.resolve(`./js.cjs`),
    },
    ts: {
        pattern: /@ts([\s\S]*?)@endts/,
        extension: `.ts`,
        loader: require.resolve(`./ts.cjs`),
    },
    vue: {
        pattern: /@vue([\s\S]*?)@endvue/,
        extension: `.vue`,
        loader: require.resolve(`./vue.cjs`),
    },
};
const make = function (entry) {
    return `import ${JSON.stringify(this.utils.contextify(this.context || this.rootContext, `${this.resource}${entry.extension}!=!${entry.loader}!${this.remainingRequest}`))};`;
};
exports.make = make;
