// @roots/bud-build/lib/handlers/items/items.js
import isBoolean from '@roots/bud-support/lodash/isBoolean';
/**
 * CSS loader
 */
export const css = ({ makeItem }) => makeItem()
    .setIdent(`css`)
    .setLoader(`css`)
    .setOptions(({ build: { rules }, hooks: { filter } }) => ({
    importLoaders: rules.css.getUse().length - 2,
    modules: false,
    sourceMap: isBoolean(filter(`build.devtool`, false))
        ? filter(`build.devtool`, false)
        : true,
}));
/**
 * CSS module loader
 */
export const cssModule = ({ makeItem }) => makeItem()
    .setIdent(`cssModule`)
    .setLoader(`css`)
    .setOptions(({ build: { rules }, hooks: { filter } }) => ({
    importLoaders: rules.cssModule.getUse().length - 2,
    modules: true,
    sourceMap: isBoolean(filter(`build.devtool`, false))
        ? filter(`build.devtool`, false)
        : true,
}));
/**
 * CSV loader
 */
export const csv = ({ makeItem }) => makeItem().setLoader(`csv`).setIdent(`csv`);
/**
 * HTML loader
 */
export const html = ({ makeItem }) => makeItem().setLoader(`html`).setIdent(`html`);
/**
 * Style loader
 */
export const style = ({ makeItem }) => makeItem().setLoader(`style`).setIdent(`style`);
/**
 * Markdown loader
 */
export const md = ({ makeItem }) => makeItem().setIdent(`md`).setLoader(`md`);
/**
 * MiniCss loader
 */
export const minicss = ({ makeItem }) => makeItem()
    .setLoader(`minicss`)
    .setIdent(`minicss`)
    .setOptions(app => ({
    publicPath: app.hooks.filter(`build.output.publicPath`),
}));
/**
 * Raw loader
 */
export const raw = ({ makeItem }) => makeItem().setLoader(`raw`).setIdent(`raw`);
/**
 * File loader
 */
export const file = ({ makeItem }) => makeItem().setLoader(`file`).setIdent(`file`);
/**
 * Xml loader
 */
export const xml = ({ makeItem }) => makeItem().setLoader(`xml`).setIdent(`xml`);
/**
 * Yml loader
 */
export const yml = ({ makeItem }) => makeItem().setLoader(`yml`).setIdent(`yml`);
