// @roots/bud-extensions/lib/extensions/html-webpack-plugin/index.js
import { __decorate } from "tslib";
import { Extension } from '@roots/bud-framework/extension';
import { disabled, label, options, } from '@roots/bud-framework/extension/decorators';
/**
 * `@roots/bud-extensions/html-webpack-plugin` adapter
 *
 * @public
 * @decorator `@label`
 * @decorator `@plugin`
 */
let BudHtmlWebpackPlugin = class BudHtmlWebpackPlugin extends Extension {
    /**
     * Make extension
     *
     * @param bud - bud instance
     * @param options - plugin options
     *
     * @public
     */
    async make(bud, options) {
        const { Plugin } = await bud.module.import(`@roots/bud-support/html-webpack-plugin`);
        return new Plugin(options);
    }
};
BudHtmlWebpackPlugin = __decorate([
    label(`@roots/bud-extensions/html-webpack-plugin`),
    options({
        filename: `index.html`,
        inject: true,
        template: `auto`,
        publicPath: app => app.publicPath(),
    }),
    disabled
], BudHtmlWebpackPlugin);
export default BudHtmlWebpackPlugin;
