// @roots/bud-framework/lib/bud.js
import { __decorate, __metadata } from "tslib";
import { bind } from '@roots/bud-support/decorators';
import isFunction from '@roots/bud-support/lodash/isFunction';
import isNull from '@roots/bud-support/lodash/isNull';
import isString from '@roots/bud-support/lodash/isString';
import isUndefined from '@roots/bud-support/lodash/isUndefined';
import { bootstrap, LIFECYCLE_EVENT_MAP } from './lifecycle/bootstrap.js';
import Value from './value.js';
/**
 * Framework abstract
 */
export class Bud {
    constructor() {
        this.services = [];
        /**
         * Value helper
         *
         * @public
         */
        this.value = Value;
    }
    isCLI() {
        return `args` in this.context;
    }
    /**
     * Compilation mode
     *
     * @remarks
     * Either `production` or `development`.
     *
     * @readonly
     * @defaultValue `production`
     * @public
     */
    get mode() {
        return this.context.mode ?? `production`;
    }
    /**
     * Name
     *
     * @readonly
     * @public
     */
    get label() {
        return this.context.label;
    }
    /**
     * Parent {@link Bud} instance
     *
     * @readonly
     * @public
     */
    get root() {
        return this.context.root ?? this;
    }
    /**
     * True when {@link Bud.mode} is `production`
     *
     * @public
     */
    get isProduction() {
        return this.mode == `production`;
    }
    /**
     * True when {@link Bud.mode} is `development`
     *
     * @public
     */
    get isDevelopment() {
        return this.mode == `development`;
    }
    /**
     * True when current instance is the parent instance
     *
     * @readonly
     * @public
     */
    get isRoot() {
        return this.root.label === this.label;
    }
    /**
     * True when current instance is a child instance
     *
     * @readonly
     * @public
     */
    get isChild() {
        return this.root.label !== this.label;
    }
    /**
     * True when child compilers
     *
     * @readonly
     * @public
     */
    get hasChildren() {
        return (!isUndefined(this.children) &&
            !isNull(this.children) &&
            Object.entries(this.children).length > 0);
    }
    /**
     * Creates a child with `bud.create` but returns the parent instance
     *
     * @public
     * @decorator `@bind`
     */
    async make(request, tap) {
        if (!this.isRoot)
            return this.fatal(new Error(`Child instances should be produced from the root context`));
        const context = isString(request)
            ? { ...this.context, label: request, root: this }
            : { ...this.context, ...request, root: this };
        if (this.isCLI() &&
            !isUndefined(this.context.args.filter) &&
            !this.context.args.filter.includes(context.label)) {
            this.log(`skipping child instance based on --filter flag:`, context.label);
            return this;
        }
        if (this.children && this.children[context.label]) {
            this.log(`returning requested child instance:`, context.label);
            return this.get(context.label);
        }
        this.log(`instantiating new bud instance`);
        const child = await new Bud().lifecycle(context);
        if (!this.children)
            this.children = { [context.label]: child };
        else
            this.children[context.label] = child;
        this.get(context.label).hooks.on(`build.dependencies`, typeof request !== `string` && request.dependsOn
            ? request.dependsOn
            : Object.values(this.children)
                .map(({ label }) => label)
                .filter(label => label !== context.label));
        if (tap)
            await tap(this.get(context.label));
        await this.get(context.label)?.api.processQueue();
        return this;
    }
    async lifecycle(context) {
        await bootstrap.bind(this)({ ...context });
        Object.entries(LIFECYCLE_EVENT_MAP).map(([eventHandle, callbackName]) => this.services
            .map(service => [service, this[service]])
            .map(([label, service]) => {
            if (!service) {
                this.error(`service not found: ${label}`, this.services);
            }
            if (!isFunction(service[callbackName]))
                return;
            this.hooks.action(eventHandle, service[callbackName].bind(service));
            this.success(`registered service callback:`, `${label}.${callbackName}`);
        }));
        await [
            `init`,
            `bootstrap`,
            `bootstrapped`,
            `register`,
            `registered`,
            `boot`,
            `booted`,
        ].reduce(async (promised, event) => {
            await promised;
            await this.hooks.fire(event);
        }, Promise.resolve());
        return this;
    }
    /**
     * Log a message
     *
     * @public
     * @decorator `@bind`
     */
    log(...messages) {
        this.context.logger.log(...messages);
        return this;
    }
    /**
     * Log an `info` level message
     *
     * @public
     * @decorator `@bind`
     */
    info(...messages) {
        this.context.logger.info(...messages);
        return this;
    }
    /**
     * Log a `success` level message
     *
     * @public
     * @decorator `@bind`
     */
    success(...messages) {
        this.context.logger.success(...messages);
        return this;
    }
    /**
     * Log a `warning` level message
     *
     * @public
     * @decorator `@bind`
     */
    warn(...messages) {
        this.context.logger.warn(...messages);
        return this;
    }
    /**
     * Log an error.
     *
     * @remarks
     * In `production` this error is treated as fatal
     * and will kill the process.
     *
     * @public
     * @decorator `@bind`
     */
    error(...messages) {
        this.context.logger.error(...messages);
        return this;
    }
    /**
     * Log and display an error.
     *
     * @remarks
     * This will always kill the process
     *
     * @public
     * @decorator `@bind`
     * @throws fatal error
     */
    fatal(error) {
        this.context.logger.fatal(error);
        return this;
    }
}
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object, Function]),
    __metadata("design:returntype", Promise)
], Bud.prototype, "make", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Promise)
], Bud.prototype, "lifecycle", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", void 0)
], Bud.prototype, "log", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", void 0)
], Bud.prototype, "info", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", void 0)
], Bud.prototype, "success", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", void 0)
], Bud.prototype, "warn", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", Bud)
], Bud.prototype, "error", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Error]),
    __metadata("design:returntype", void 0)
], Bud.prototype, "fatal", null);
