// @roots/bud-framework/lib/services/console.js
import { __decorate, __metadata } from "tslib";
import { bind } from '@roots/bud-support/decorators';
import patchConsole from '@roots/bud-support/patch-console';
import { Service } from '../service.js';
/**
 * ConsoleBuffer service class
 *
 * @remarks
 * Intercepts console function calls and emits them using the bud logger.
 * Deduplicates and trims console output.
 */
export default class ConsoleBuffer extends Service {
    constructor() {
        super(...arguments);
        /**
         * Received messages
         * @public
         */
        this.messages = [];
    }
    fetchAndRemove() {
        const messages = [...this.messages];
        this.messages = [];
        return messages;
    }
    /**
     * `init` callback
     *
     * @public
     * @decorator `@bind`
     */
    async init(bud) {
        if (!bud.isCLI() || (bud.isCLI() && bud.context.args?.ci))
            return;
        // Patch the console, and assign the restore function
        this.restore = patchConsole((stream, data) => {
            if (!data || data === ``)
                return;
            const message = data.trim();
            // Ignore empty
            if (!message)
                return;
            // Ignore messages that have been logged before
            if (this.messages.some(stale => stale.message === message && stale.stream === stream))
                return;
            // Add message to buffer
            this.messages.push({ stream, message });
        });
    }
}
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Function]),
    __metadata("design:returntype", Promise)
], ConsoleBuffer.prototype, "init", null);
