// @roots/bud-server/lib/server/base.js
import { __decorate, __metadata } from "tslib";
import { bind } from '@roots/bud-support/decorators';
/**
 * Node server
 *
 * @remarks
 * Base class. Extended by either `http` or `https` class.
 */
export class BaseServer {
    /**
     * Logger
     *
     * @public
     */
    get logger() {
        return this.app.context.logger.scope(this.constructor.name.toLowerCase());
    }
    /**
     * Options
     *
     * @public
     */
    get options() {
        return this.app.hooks.filter(`dev.options`, {});
    }
    /**
     * Constructor
     *
     * @param app - Bud
     * @public
     */
    constructor(app) {
        this.app = app;
    }
    /**
     * Listen
     *
     * @public
     * @decorator `@bind`
     */
    async listen() {
        this.instance
            .listen(this.app.hooks.filter(`dev.listenOptions`, {
            host: this.app.server.url.hostname,
            port: Number(this.app.server.url.port),
        }))
            .on(`listening`, this.app.hooks.filter(`dev.onListening`, this.onListening))
            .on(`request`, this.app.hooks.filter(`dev.onRequest`, this.onRequest))
            .on(`error`, this.app.hooks.filter(`dev.onError`, this.onError));
    }
    /**
     * Server listen event
     *
     * @public
     * @decorator `@bind`
     */
    onListening(...param) {
        this.logger.info(`listening`, ...param);
    }
    /**
     * Server request event
     *
     * @public
     * @decorator `@bind`
     */
    async onRequest(request, response) {
        this.logger.log(`[${response.statusCode}]`, request.url, response.statusMessage ?? ``);
        return response;
    }
    /**
     * Server error event
     *
     * @public
     * @decorator `@bind`
     */
    onError(error) {
        error.name = `bud.js server error`;
        error.message = error?.message ?? error.toString();
        throw error;
    }
}
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", Promise)
], BaseServer.prototype, "listen", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Object]),
    __metadata("design:returntype", void 0)
], BaseServer.prototype, "onListening", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Function, Function]),
    __metadata("design:returntype", Promise)
], BaseServer.prototype, "onRequest", null);
__decorate([
    bind,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [Error]),
    __metadata("design:returntype", void 0)
], BaseServer.prototype, "onError", null);
